### PR TITLE
Update role assignment cmdlet

### PR DESCRIPTION
**Pull request by dagoroz**
## Description

Added cmdlet "Set-AzRoleAssignment" and small suggestions that arose from the design review

## Checklist

- [x] I have read the [_Submitting Changes_](../blob/master/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](../blob/master/CONTRIBUTING.md)
- [x] The title of the PR is clear and informative
- [x] The appropriate `ChangeLog.md` file(s) has been updated:
    - For any service, the `ChangeLog.md` file can be found at `src/{{SERVICE}}/{{SERVICE}}/ChangeLog.md`
    - A snippet outlining the change(s) made in the PR should be written under the `## Upcoming Release` header -- no new version header should be added
- [x] The PR does not introduce [breaking changes](../blob/master/documentation/breaking-changes/breaking-changes-definition.md)
- x] If applicable, the changes made in the PR have proper test coverage
- [x] For public API changes to cmdlets:
    - [x] a cmdlet design review was approved for the changes in [this repository](https://github.com/Azure/azure-powershell-cmdlet-review-pr) (_Microsoft internal only_)
    - [x] the markdown help files have been regenerated using the commands listed [here](../blob/master/documentation/development-docs/help-generation.md#updating-all-markdown-files-in-a-module)

## Link to review
https://github.com/Azure/azure-powershell-cmdlet-review-pr/issues/689

### CODE REVIEW

**Comment by adxsdkps** [comment]
Can one of the admins verify this patch?

File: src/Resources/Resources/RoleAssignments/SetAzureRoleAssignmentCommand.cs [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+            if (!isValidRequest)
+            {
+                throw new ArgumentException("Changing a property other than 'Description', 'Condition' or 'Condition Version' is currently not supported.");
+            }
+
+            // If ConditionVersion is changed, validate it's in the allowed values
+            var oldConditionVersion = double.Parse(InputObject.ConditionVersion);
+            var newConditionVersion = double.Parse(fetchedRole.ConditionVersion);
+
+            // A condition version can change but currently we don't suport downgrading to 1.0
```
Comment (jiasli) [inline src/Resources/Resources/RoleAssignments/SetAzureRoleAssignmentCommand.cs:114]:
```suggestion
            // A condition version can change but currently we don't support downgrading to 1.0
```

File: src/Resources/Resources/RoleAssignments/SetAzureRoleAssignmentCommand.cs [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+            // we only verify the change if it's a downgrade
+            if ((oldConditionVersion > newConditionVersion) && (newConditionVersion < 2.0))
+            {
+                throw new ArgumentException("Condition version diferent than '2.0' is not supported for update operations");
```
Comment (jiasli) [inline src/Resources/Resources/RoleAssignments/SetAzureRoleAssignmentCommand.cs:118]:
```suggestion
                throw new ArgumentException("Condition version difference than '2.0' is not supported for update operations");
```

File: src/Resources/Resources/help/Az.Resources.md [hunk 3]
```diff
@@ -341,6 +341,9 @@ Modifies a resource group.
 ### [Set-AzResourceLock](Set-AzResourceLock.md)
 Modifies a resource lock.
 
+### [Set-AzRoleAssignment](Set-AzRoleAssignment.md)
+Update a role assignment
```
Comment (jiasli) [inline src/Resources/Resources/help/Az.Resources.md:345]:
```suggestion
Update a role assignment.
```

**Comment by jiasli** [comment]
Could you update the [New-AzRoleAssignment.md](https://github.com/Azure/azure-powershell/blob/master/src/Resources/Resources/help/New-AzRoleAssignment.md) and also add [Set-AzRoleAssignment.md](https://github.com/Azure/azure-powershell/blob/master/src/Resources/Resources/help/Set-AzRoleAssignment.md) to provide more examples on the updated commands?

File: src/Resources/Resources/ChangeLog.md [hunk 4]
```diff
@@ -18,6 +18,7 @@
         - Additional information about change #1
 -->
 ## Upcoming Release
+* Created new CMDlet 'Set-RoleAssignment'
```
Comment (isra-fel) [inline src/Resources/Resources/ChangeLog.md:21]:
```suggestion
* Created new cmdlet `Set-AzRoleAssignment`
```

File: src/Resources/Resources/RoleAssignments/NewAzureRoleAssignmentCommand.cs [hunk 5]
```diff
@@ -221,6 +229,34 @@ public class NewAzureRoleAssignmentCommand : ResourcesBaseCmdlet
...
+                    Guid.TryParse(RoleAssignment.RoleDefinitionId,out guid);
+                    this.RoleDefinitionId = guid;
+                    this.Description = RoleAssignment.Description;
+                    this.Condition = RoleAssignment.Condition;
+                    this.ConditionVersion = RoleAssignment.ConditionVersion;
+                }
+                catch (JsonException)
+                {
+                    WriteVerbose("Deserializing the input role definition failed.");
+                    throw;
```
Comment (isra-fel) [inline src/Resources/Resources/RoleAssignments/NewAzureRoleAssignmentCommand.cs:257]: I suppose we should throw meaningful error message to the user.
A verbose log is by default not displayed.

File: src/Resources/Resources/RoleAssignments/SetAzureRoleAssignmentCommand.cs [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+                }
+
+                try
+                {
+                    InputObject = JsonConvert.DeserializeObject<PSRoleAssignment>(File.ReadAllText(fileName));
+                }
+                catch (JsonException)
+                {
+                    WriteVerbose("Deserializing the input role definition failed.");
+                    throw;
```
Comment (isra-fel) [inline src/Resources/Resources/RoleAssignments/SetAzureRoleAssignmentCommand.cs:64]: Same here.

File: src/Resources/Resources/RoleAssignments/SetAzureRoleAssignmentCommand.cs [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+using System.IO;
+using System.Linq;
+using System.Management.Automation;
+
+namespace Microsoft.Azure.Commands.Resources
+{
+    /// <summary>
+    /// Updates an existing role assignment.
+    /// </summary>
+    [Cmdlet("Set", ResourceManager.Common.AzureRMConstants.AzureRMPrefix + "RoleAssignment", DefaultParameterSetName = ParameterSet.RoleAssignment), OutputType(typeof(PSRoleAssignment))]
```
Comment (isra-fel) [inline src/Resources/Resources/RoleAssignments/SetAzureRoleAssignmentCommand.cs:32]: Sorry I forgot to mention in the design review that all the cmdlets that make change on azure should support "ShouldProcess", which basically means you should declare `SupportShouldProcess` in the attribute and call `ConfirmAction()` in `ExecuteCmdlet()`. You can take `Remove-AzRoleAssignment` as an example: https://github.com/Azure/azure-powershell/blob/fee5a593ce9437f8796de39f25ecce1cad1115c1/src/Resources/Resources/RoleAssignments/RemoveAzureRoleAssignmentCommand.cs

**Comment by isra-fel** [comment]
Hi @dagoroz , only thing left is to regenerate help docs: https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/help-generation.md#updating-help-after-making-cmdlet-changes

Some of the changes were not reflected in docs, for example there should be description of `-WhatIf` parameter in `Set-AzRoleAssignment.md`

**Comment by dagoroz** [comment]
@isra-fel I have addressed all comments and all checks are passing now
Would you kindly approve?

**Pull request merged**